pkg/agent/storage/storage.go

```diff
@@ -0,0 +1,287 @@
+package storage
+
+import (
+	"crypto/x509"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"io"
+	"io/fs"
+	"os"
+	"path/filepath"
+	"sync"
+	"time"
+
+	"github.com/spiffe/spire/pkg/common/diskutil"
+	"github.com/spiffe/spire/pkg/common/pemutil"
+)
+
+var (
+	ErrNotCached = errors.New("not cached")
+)
+
+type Storage interface {
+	// LoadSVID loads the SVID from storage. Returns ErrNotCached if the SVID
+	// does not exist in the cache.
+	LoadSVID() ([]*x509.Certificate, error)
+
+	// StoreSVID stores the SVID.
+	StoreSVID(certs []*x509.Certificate) error
+
+	// DeleteSVID deletes the SVID.
+	DeleteSVID() error
+
+	// LoadBundle loads the SVID from storage. Returns ErrNotCached if the
+	// bundle does not exist in the cache.
+	LoadBundle() ([]*x509.Certificate, error)
+
+	// StoreBundle stores the bundle.
+	StoreBundle(certs []*x509.Certificate) error
+}
+
+func Open(dir string) (Storage, error) {
+	legacySVID, legacySVIDTime, err := loadLegacySVID(dir)
+	if err != nil && !errors.Is(err, fs.ErrNotExist) {
+		return nil, err
+	}
+
+	legacyBundle, legacyBundleTime, err := loadLegacyBundle(dir)
+	if err != nil && !errors.Is(err, fs.ErrNotExist) {
+		return nil, err
+	}
+
+	data, dataTime, err := loadData(dir)
+	if err != nil && !errors.Is(err, fs.ErrNotExist) {
+		return nil, err
+	}
+
+	storeNow := false
+
+	if !legacySVIDTime.IsZero() && (dataTime.IsZero() || dataTime.Before(legacySVIDTime)) {
+		storeNow = true
+		data.SVID = legacySVID
+	}
+
+	if !legacyBundleTime.IsZero() && (dataTime.IsZero() || dataTime.Before(legacyBundleTime)) {
+		storeNow = true
+		data.Bundle = legacyBundle
+	}
+
+	if storeNow {
+		if err := storeData(dir, data); err != nil {
+			return nil, err
+		}
+	}
+
+	return &storage{
+		dir:  dir,
+		data: data,
+	}, nil
+}
+
+type storage struct {
+	dir string
+
+	mtx  sync.RWMutex
+	data storageData
+}
+
+func (s *storage) LoadBundle() ([]*x509.Certificate, error) {
+	s.mtx.RLock()
+	defer s.mtx.RUnlock()
+
+	if len(s.data.Bundle) == 0 {
+		return nil, ErrNotCached
+	}
+	return s.data.Bundle, nil
+}
+
+func (s *storage) StoreBundle(bundle []*x509.Certificate) error {
+	s.mtx.Lock()
+	defer s.mtx.Unlock()
+
+	if err := storeLegacyBundle(s.dir, bundle); err != nil {
+		return err
+	}
+
+	data := s.data
+	data.Bundle = bundle
+
+	if err := storeData(s.dir, data); err != nil {
+		return err
+	}
+
+	s.data = data
+	return nil
+}
+
+func (s *storage) LoadSVID() ([]*x509.Certificate, error) {
+	s.mtx.RLock()
+	defer s.mtx.RUnlock()
+
+	if len(s.data.SVID) == 0 {
+		return nil, ErrNotCached
+	}
+	return s.data.SVID, nil
+}
+
+func (s *storage) StoreSVID(svid []*x509.Certificate) error {
+	s.mtx.Lock()
+	defer s.mtx.Unlock()
+
+	if err := storeLegacySVID(s.dir, svid); err != nil {
```

Comment:
Should we add a "TODO" comment to remove this after a deprecation cycle and maybe file an issue with the target version when we can safely remove the legacy stuff?

Reply:
Oh yes. Thanks, will add.